File: Dynamic Programming/Coin Change Problem/Coin Change(Number of ways getting sum).java
import java.io.*;
import java.util.*;

public class CoinChange {
    public static long getWays(int n, int sum, int[] coins) {
        long[][] dp = new long[n][sum+1];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j <= sum; j++){
                if(j == 0){
                    dp[i][j] = 1;
                } else {
                    long excluding = 0;
                    long including = 0;
                    
                    if(i > 0){     				//excluding coins[i]
                        excluding  = dp[i-1][j];
                    }
                    if(j >= coins[i]){          //including coins[i]
                        // coins[i] can pick multiple times
                        // if want to pick exactly one time then use dp[i-1][j-coins[i]]              
                        including = dp[i][j-coins[i]];  
                    }
                    
                    dp[i][j] = excluding + including;
                }
                
            }
        }
        
        return dp[n-1][sum];
    }

    public static void main(String[] args) throws IOException {
        Scanner scan = new Scanner(System.in);
        
        int sum = scan.nextInt();
        int n = scan.nextInt();
        int[] coins = new int[n];
        
        for(int i = 0; i  < n; i++){
            coins[i] = scan.nextInt();
        }
        
        System.out.println(getWays(n, sum, coins));
    }
}



// linear space [can take same element multiple times]
#include <bits/stdc++.h>
using namespace std;

int countWays(int arr[], int m, int N)
{
	int count[N + 1];
	memset(count, 0, sizeof(count));
	
	// base case
	count[0] = 1;
	
	// count ways for all values up
	// to 'N' and store the result
	for (int i = 1; i <= N; i++)
		for (int j = 0; j < m; j++)
			if (i >= arr[j])
				count[i] += count[i - arr[j]];
	
	return count[N];
	
}


int main()
{
	int arr[] = {1, 5, 6};
	int m = sizeof(arr) / sizeof(arr[0]);
	int N = 7;
	cout << "Total number of ways = "
		<< countWays(arr, m, N);
	return 0;
}

File: Dynamic Programming/Coin Change Problem/coin_change(number of ways getting sum).cpp
#include<bits/stdc++.h>
using namespace std;
#define endl    '\n'
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<vector<int>> vii;
 
/*****User defined function*****/
void solve_problem() {
    int n, sum;
    cin >> n >> sum;

    vector<int> a(n);
    for (int i = 0; i < n; i++)
        cin >> a[i];

    vector<vector<ll>> dp(n, vector<ll>(sum+1, 0));
    for(int i = 0; i < n; i++)
        dp[i][0] = 1;


    for (int i = 0; i < n; i++) {
        for(int j = 1; j <= sum; j++) {
            int including = 0, excluding = 0;
            // including a[i] 
            // a[i] can multiple times, if use exactly one time then dp[i-1][j-a[i]]
            if(j >= a[i])
                including = dp[i][j-a[i]];
            if(i > 0)
                excluding = dp[i-1][j];
            dp[i][j] = including + excluding;
        }
    }

    cout << "Maximum number of way: " << dp[n-1][sum] << endl;
} 
 
/*****main function*****/
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif 
 
    cout << fixed << showpoint;
    cout << setprecision(5);
 
    int test = 1;
    // cin >> test;
    while(test--) {
        solve_problem();
    }
 
    return 0;
} 